File: app/src/main/java/md/fusionworks/lifehack/ui/NavigationDrawerActivity.java
package md.fusionworks.lifehack.ui;

import android.content.res.Configuration;
import android.os.Bundle;
import android.os.Handler;
import android.support.v4.view.GravityCompat;
import android.support.v4.widget.DrawerLayout;
import android.support.v7.app.ActionBarDrawerToggle;
import android.support.v7.widget.Toolbar;
import android.view.View;
import android.view.ViewGroup;
import android.widget.ImageView;
import android.widget.TextView;
import butterknife.Bind;
import butterknife.ButterKnife;
import java.util.ArrayList;
import md.fusionworks.lifehack.R;
import md.fusionworks.lifehack.util.Constant;
import md.fusionworks.lifehack.util.UIUtil;

/**
 * Created by admin on 03.09.2015.
 */
public class NavigationDrawerActivity extends BaseActivity {

  protected static final int DRAWER_ITEM_INVALID = -1;
  protected static final int DRAWER_ITEM_SEPARATOR = -2;

  private static final int DRAWER_LAUNCH_DELAY = 250;

  private static final int[] DRAWER_TITLE_RES_ID = new int[] {
      R.string.drawer_item_main, R.string.drawer_item_life_hacks,
      R.string.drawer_item_exchange_rates, R.string.drawer_item_life_movies,
      R.string.drawer_item_life_taxi, R.string.drawer_item_prices, R.string.drawer_item_sales,
      R.string.drawer_item_life_poster
      // R.string.drawer_item_settings
  };

  private static final int[] DRAWER_ICON_RES_ID = new int[] {
      R.drawable.ic_home, R.drawable.ic_newspaper, R.drawable.ic_currency_exchange,
      R.drawable.ic_movie, R.drawable.ic_taxi, R.drawable.ic_price_tag, R.drawable.ic_sale,
      R.drawable.ic_advertising,
      //R.drawable.ic_lens_black_24dp,
  };

  private static final boolean[] DRAWER_ITEM_ACTIVE_STATE = new boolean[] {
      true, true, true, false, true, false, true, false,
      //false
  };

  @Bind(R.id.toolbar) Toolbar toolbar;
  @Bind(R.id.drawerLayout) DrawerLayout drawerLayout;
  @Bind(R.id.drawerItemsContainer) ViewGroup drawerItemsListContainer;

  private ArrayList<Integer> drawerItems = new ArrayList<Integer>();
  private View[] drawerItemViews = null;
  private ActionBarDrawerToggle actionBarDrawerToggle;
  private Handler handler;

  @Override protected void onCreate(Bundle savedInstanceState) {
    super.onCreate(savedInstanceState);
    handler = new Handler();
  }

  @Override public void onPostCreate(Bundle savedInstanceState) {
    super.onPostCreate(savedInstanceState);

    ButterKnife.bind(this);

    setSupportActionBar(toolbar);
    getSupportActionBar().setDisplayHomeAsUpEnabled(true);

    setupDrawerLayout();
  }

  private void setupDrawerLayout() {
    if (drawerLayout == null) {
      return;
    }

    actionBarDrawerToggle =
        new ActionBarDrawerToggle(this, drawerLayout, R.string.app_name, R.string.app_name);
    actionBarDrawerToggle.setDrawerIndicatorEnabled(true);

    drawerLayout.setDrawerListener(actionBarDrawerToggle);
    toolbar.setNavigationOnClickListener(view -> drawerLayout.openDrawer(GravityCompat.START));
    actionBarDrawerToggle.syncState();

    populateDrawerItems();
  }

  @Override public void onConfigurationChanged(Configuration newConfig) {
    super.onConfigurationChanged(newConfig);
    actionBarDrawerToggle.onConfigurationChanged(newConfig);
  }

  public void setTitle(String title) {
    getSupportActionBar().setTitle(title);
  }

  private void populateDrawerItems() {
    drawerItems.add(Constant.DRAWER_ITEM_MAIN);
    drawerItems.add(Constant.DRAWER_ITEM_LIFE_HACKS);
    drawerItems.add(Constant.DRAWER_ITEM_EXCHANGE_RATES);
    drawerItems.add(Constant.DRAWER_ITEM_SALES);
    drawerItems.add(Constant.DRAWER_ITEM_TAXI);
    drawerItems.add(Constant.DRAWER_ITEM_PRICES);
    drawerItems.add(Constant.DRAWER_ITEM_MOVIES);
    drawerItems.add(Constant.DRAWER_ITEM_POSTER);
    // drawerItems.add(Constant.DRAWER_ITEM_SETTINGS);

    createDrawerItems();
  }

  private void createDrawerItems() {
    if (drawerItemsListContainer == null) {
      return;
    }

    drawerItemViews = new View[drawerItems.size()];
    drawerItemsListContainer.removeAllViews();
    int i = 0;
    for (int itemId : drawerItems) {
      drawerItemViews[i] = makeDrawerItem(itemId, drawerItemsListContainer);
      drawerItemsListContainer.addView(drawerItemViews[i]);
      ++i;
    }
  }

  private View makeDrawerItem(final int itemId, ViewGroup container) {
    boolean selected = getSelfDrawerItem() == itemId;
    int layoutToInflate = 0;
    if (itemId == DRAWER_ITEM_SEPARATOR) {
      layoutToInflate = R.layout.navdrawer_separator;
    } else {
      layoutToInflate = R.layout.navdrawer_item;
    }
    View view = getLayoutInflater().inflate(layoutToInflate, container, false);

    if (isSeparator(itemId)) {

      UIUtil.setAccessibilityIgnore(view);
      return view;
    }

    ImageView iconView = (ImageView) view.findViewById(R.id.icon);
    TextView titleView = (TextView) view.findViewById(R.id.title);
    int iconId = itemId >= 0 && itemId < DRAWER_ICON_RES_ID.length ? DRAWER_ICON_RES_ID[itemId] : 0;
    int titleId =
        itemId >= 0 && itemId < DRAWER_TITLE_RES_ID.length ? DRAWER_TITLE_RES_ID[itemId] : 0;
    boolean isActive =
        itemId >= 0 && itemId < DRAWER_ITEM_ACTIVE_STATE.length ? DRAWER_ITEM_ACTIVE_STATE[itemId]
            : false;

    iconView.setVisibility(iconId > 0 ? View.VISIBLE : View.GONE);
    if (iconId > 0) {
      iconView.setImageResource(iconId);
    }
    titleView.setText(getString(titleId));

    formatDrawerItem(view, itemId, selected, isActive);

    if (isActive) view.setOnClickListener(v -> onDrawerItemClicked(itemId));

    return view;
  }

  public int getSelfDrawerItem() {
    return DRAWER_ITEM_INVALID;
  }

  private boolean isSeparator(int itemId) {
    return itemId == DRAWER_ITEM_SEPARATOR;
  }

  private boolean isSimpleActivity(int itemId) {
    return itemId == DRAWER_ITEM_INVALID;
  }

  private void formatDrawerItem(View view, int itemId, boolean selected, boolean isActive) {
    if (isSeparator(itemId)) {
      // not applicable
      return;
    }

    ImageView iconView = (ImageView) view.findViewById(R.id.icon);
    TextView titleView = (TextView) view.findViewById(R.id.title);

    if (selected) {
      view.setBackgroundResource(R.drawable.selected_navdrawer_item_background);
    }

    int titleColorRes;
    int iconColorRes;

    if (isActive) {

      titleColorRes =
          selected ? R.color.navdrawer_text_color_selected : R.color.navdrawer_text_color;
      iconColorRes = selected ? R.color.navdrawer_icon_tint_selected : R.color.navdrawer_icon_tint;
    } else {

      titleColorRes = R.color.navdrawer_text_color_inactive;
      iconColorRes = R.color.navdrawer_icon_tint_inactive;
    }

    titleView.setTextColor(getResources().getColor(titleColorRes));
    iconView.setColorFilter(getResources().getColor(iconColorRes));
  }

  private void onDrawerItemClicked(final int itemId) {
    if (itemId == getSelfDrawerItem()) {
      drawerLayout.closeDrawer(GravityCompat.START);
      return;
    }

    if (isSimpleActivity(itemId)) {
      goToDrawerItem(itemId);
    } else {

      handler.postDelayed(() -> goToDrawerItem(itemId), DRAWER_LAUNCH_DELAY);

      setSelectedDrawerItem(itemId);
    }

    drawerLayout.closeDrawer(GravityCompat.START);
  }

  private void setSelectedDrawerItem(int itemId) {
    if (drawerItemViews != null) {
      for (int i = 0; i < drawerItemViews.length; i++) {
        if (i < drawerItems.size()) {

          int thisItemId = drawerItems.get(i);
          boolean isActive = itemId >= 0 && itemId < DRAWER_ITEM_ACTIVE_STATE.length
              ? DRAWER_ITEM_ACTIVE_STATE[itemId] : false;
          formatDrawerItem(drawerItemViews[i], thisItemId, itemId == thisItemId, isActive);
        }
      }
    }
  }

  public void goToDrawerItem(int item) {
    switch (item) {
      case Constant.DRAWER_ITEM_MAIN:
        getNavigator().navigateToMainActivity(this);
        finish();
        break;
      case Constant.DRAWER_ITEM_EXCHANGE_RATES:
        getNavigator().navigateToExchangeRatesActivity(this);
        finish();
        break;
      case Constant.DRAWER_ITEM_TAXI:
        getNavigator().navigateToTaxiActivity(this);
        finish();
        break;
      case Constant.DRAWER_ITEM_LIFE_HACKS:
        getNavigator().navigateToLifeHacksActivity(this);
        finish();
        break;
      case Constant.DRAWER_ITEM_SALES:
        getNavigator().navigateToSalesActivity(this);
        finish();
        break;
    }
  }
}

File: app/src/main/java/md/fusionworks/lifehack/ui/NavigationDrawerActivity.kt
package md.fusionworks.lifehack.ui

import android.content.res.Configuration
import android.os.Bundle
import android.os.Handler
import android.support.v4.view.GravityCompat
import android.support.v4.widget.DrawerLayout
import android.support.v7.app.ActionBarDrawerToggle
import android.support.v7.widget.AppCompatButton
import android.support.v7.widget.Toolbar
import android.view.View
import android.view.ViewGroup
import android.widget.ImageView
import android.widget.TextView
import md.fusionworks.lifehack.R
import md.fusionworks.lifehack.ui.base.view.BaseActivity
import md.fusionworks.lifehack.util.Constant
import md.fusionworks.lifehack.util.LocaleHelper
import md.fusionworks.lifehack.util.setAccessibilityIgnore
import org.jetbrains.anko.find
import java.util.*

/**
 * Created by admin on 03.09.2015.
 */
open class NavigationDrawerActivity : BaseActivity() {

  private val toolbar by lazy { find<Toolbar>(R.id.toolbar) }
  private val drawerLayout by lazy { find<DrawerLayout>(R.id.drawerLayout) }
  private val drawerItemsListContainer by lazy { find<ViewGroup>(R.id.drawerItemsContainer) }
  private val languageButton by lazy { find<AppCompatButton>(R.id.languageButton) }

  private val drawerItems = ArrayList<Int>()
  private lateinit var drawerItemViews: MutableList<View>
  private var actionBarDrawerToggle: ActionBarDrawerToggle? = null
  private var handler: Handler? = null

  override fun onCreate(savedInstanceState: Bundle?) {
    LocaleHelper.onCreate(this)
    super.onCreate(savedInstanceState)
    handler = Handler()
  }

  public override fun onPostCreate(savedInstanceState: Bundle?) {
    super.onPostCreate(savedInstanceState)

    setSupportActionBar(toolbar)
    supportActionBar!!.setDisplayHomeAsUpEnabled(true)

    setupDrawerLayout()

    languageButton.setOnClickListener {
      val currentLanguage = LocaleHelper.getLanguage(this)
      val toggleLanguage = if (currentLanguage == Constant.LANG_RU) Constant.LANG_RO else Constant.LANG_RU
      LocaleHelper.setLocale(this, toggleLanguage)

      drawerLayout!!.closeDrawer(GravityCompat.START)
      handler!!.postDelayed({ recreate() }, DRAWER_LAUNCH_DELAY.toLong())
    }
  }

  private fun setupDrawerLayout() {
    if (drawerLayout == null) {
      return
    }

    actionBarDrawerToggle = ActionBarDrawerToggle(this, drawerLayout, R.string.app_name,
        R.string.app_name)
    actionBarDrawerToggle!!.isDrawerIndicatorEnabled = true

    drawerLayout!!.setDrawerListener(actionBarDrawerToggle)
    toolbar.setNavigationOnClickListener { view -> drawerLayout!!.openDrawer(GravityCompat.START) }
    actionBarDrawerToggle!!.syncState()

    populateDrawerItems()
  }

  override fun onConfigurationChanged(newConfig: Configuration) {
    super.onConfigurationChanged(newConfig)
    actionBarDrawerToggle!!.onConfigurationChanged(newConfig)
  }

  fun setTitle(title: String) {
    supportActionBar!!.title = title
  }

  private fun populateDrawerItems() {
    drawerItems.add(Constant.DRAWER_ITEM_MAIN)
    drawerItems.add(Constant.DRAWER_ITEM_LIFE_HACKS)
    drawerItems.add(Constant.DRAWER_ITEM_EXCHANGE_RATES)
    drawerItems.add(Constant.DRAWER_ITEM_SALES)
    drawerItems.add(Constant.DRAWER_ITEM_TAXI)
    drawerItems.add(Constant.DRAWER_ITEM_MOVIES)
    drawerItems.add(Constant.DRAWER_ITEM_PRICES)
    drawerItems.add(Constant.DRAWER_ITEM_POSTER)
    // drawerItems.add(Constant.DRAWER_ITEM_SETTINGS);

    createDrawerItems()
  }

  private fun createDrawerItems() {

    drawerItemsListContainer.let {
      drawerItemViews = arrayListOf<View>()
      drawerItemsListContainer!!.removeAllViews()
      var i = 0
      for (itemId in drawerItems) {
        drawerItemViews.add(makeDrawerItem(itemId, drawerItemsListContainer))
        drawerItemsListContainer!!.addView(drawerItemViews!![i])
        ++i
      }
    }
  }

  private fun makeDrawerItem(itemId: Int, container: ViewGroup): View {
    val selected = getSelfDrawerItem() == itemId
    var layoutToInflate = 0
    if (itemId == DRAWER_ITEM_SEPARATOR) {
      layoutToInflate = R.layout.navdrawer_separator
    } else {
      layoutToInflate = R.layout.navdrawer_item
    }
    val view = layoutInflater.inflate(layoutToInflate, container, false)

    if (isSeparator(itemId)) {

      view.setAccessibilityIgnore()
      return view
    }

    val iconView = view.findViewById(R.id.icon) as ImageView
    val titleView = view.findViewById(R.id.title) as TextView
    val iconId = if (itemId >= 0 && itemId < DRAWER_ICON_RES_ID.size) DRAWER_ICON_RES_ID[itemId] else 0
    val titleId = if (itemId >= 0 && itemId < DRAWER_TITLE_RES_ID.size) DRAWER_TITLE_RES_ID[itemId] else 0
    val isActive = if (itemId >= 0 && itemId < DRAWER_ITEM_ACTIVE_STATE.size) DRAWER_ITEM_ACTIVE_STATE[itemId]
    else false

    iconView.visibility = if (iconId > 0) View.VISIBLE else View.GONE
    if (iconId > 0) {
      iconView.setImageResource(iconId)
    }
    titleView.text = getString(titleId)

    formatDrawerItem(view, itemId, selected, isActive)

    if (isActive) view.setOnClickListener { v -> onDrawerItemClicked(itemId) }

    return view
  }

  open fun getSelfDrawerItem() = DRAWER_ITEM_INVALID

  private fun isSeparator(itemId: Int): Boolean {
    return itemId == DRAWER_ITEM_SEPARATOR
  }

  private fun isSimpleActivity(itemId: Int): Boolean {
    return itemId == DRAWER_ITEM_INVALID
  }

  private fun formatDrawerItem(view: View, itemId: Int, selected: Boolean, isActive: Boolean) {
    if (isSeparator(itemId)) {
      // not applicable
      return
    }

    val iconView = view.findViewById(R.id.icon) as ImageView
    val titleView = view.findViewById(R.id.title) as TextView

    if (selected) {
      view.setBackgroundResource(R.drawable.selected_navdrawer_item_background)
    }

    val titleColorRes: Int
    val iconColorRes: Int

    if (isActive) {

      titleColorRes = if (selected) R.color.navdrawer_text_color_selected else R.color.navdrawer_text_color
      iconColorRes = if (selected) R.color.navdrawer_icon_tint_selected else R.color.navdrawer_icon_tint
    } else {

      titleColorRes = R.color.navdrawer_text_color_inactive
      iconColorRes = R.color.navdrawer_icon_tint_inactive
    }

    titleView.setTextColor(resources.getColor(titleColorRes))
    iconView.setColorFilter(resources.getColor(iconColorRes))
  }

  private fun onDrawerItemClicked(itemId: Int) {
    if (itemId == getSelfDrawerItem()) {
      drawerLayout!!.closeDrawer(GravityCompat.START)
      return
    }

    if (isSimpleActivity(itemId)) {
      goToDrawerItem(itemId)
    } else {

      handler!!.postDelayed({ goToDrawerItem(itemId) }, DRAWER_LAUNCH_DELAY.toLong())

      setSelectedDrawerItem(itemId)
    }

    drawerLayout!!.closeDrawer(GravityCompat.START)
  }

  private fun setSelectedDrawerItem(itemId: Int) {
    if (drawerItemViews != null) {
      for (i in drawerItemViews!!.indices) {
        if (i < drawerItems.size) {

          val thisItemId = drawerItems[i]
          val isActive = if (itemId >= 0 && itemId < DRAWER_ITEM_ACTIVE_STATE.size) DRAWER_ITEM_ACTIVE_STATE[itemId]
          else false
          formatDrawerItem(drawerItemViews!![i], thisItemId, itemId == thisItemId, isActive)
        }
      }
    }
  }

  fun goToDrawerItem(item: Int) {
    when (item) {
      Constant.DRAWER_ITEM_MAIN -> {
        navigator.navigateToMainActivity(this)
        finish()
      }
      Constant.DRAWER_ITEM_EXCHANGE_RATES -> {
        navigator.navigateToExchangeRatesActivity(this)
        finish()
      }
      Constant.DRAWER_ITEM_TAXI -> {
        navigator.navigateToTaxiActivity(this)
        finish()
      }
      Constant.DRAWER_ITEM_LIFE_HACKS -> {
        navigator.navigateToLifeHacksActivity(this)
        finish()
      }
      Constant.DRAWER_ITEM_SALES -> {
        navigator.navigateToSalesActivity(this)
        finish()
      }
      Constant.DRAWER_ITEM_MOVIES -> {
        navigator.navigateToMoviesActivity(this)
        finish()
      }
    }
  }

  companion object {

    protected val DRAWER_ITEM_INVALID = -1
    protected val DRAWER_ITEM_SEPARATOR = -2

    private val DRAWER_LAUNCH_DELAY = 250

    private val DRAWER_TITLE_RES_ID = intArrayOf(R.string.drawer_item_main,
        R.string.drawer_item_life_hacks, R.string.drawer_item_exchange_rates,
        R.string.drawer_item_movies, R.string.drawer_item_taxi, R.string.drawer_item_prices,
        R.string.drawer_item_sales, R.string.drawer_item_poster)

    private val DRAWER_ICON_RES_ID = intArrayOf(R.drawable.ic_home, R.drawable.ic_newspaper,
        R.drawable.ic_currency_exchange, R.drawable.ic_movie, R.drawable.ic_taxi,
        R.drawable.ic_price_tag, R.drawable.ic_sale,
        R.drawable.ic_advertising)

    private val DRAWER_ITEM_ACTIVE_STATE = booleanArrayOf(true, true, true, true, true, false, true,
        false)
  }
}
